### PR TITLE
vivaldi: allow selecting qt5 or 6 and gtk3 or 4

### PR DESCRIPTION
**Pull request by matklad**
## Description of changes

Make vivaldi's qt and gtk versions overridable. qt5 and gtk3 are still used by default, but the user now have the option of `vivaldi.override { qt = qt6; }`

Notably, overriding qt to version 6 is required to make vivaldi work with plasma 6:

https://github.com/NixOS/nixpkgs/pull/286522#issuecomment-1960433519

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by NickCao** [comment]
> Notably, overriding qt to version 6 is required to make vivaldi work with plasma 6:

I don't think so, please try using `qt5.wrapQtAppsHook`

**Comment by matklad** [comment]
> > Notably, overriding qt to version 6 is required to make vivaldi work with plasma 6:
> 
> I don't think so,

It definitely does not work on my box without that override

> please try using `qt5.wrapQtAppsHook`

I gather we intentionally avoid the wrapper here?

https://github.com/NixOS/nixpkgs/pull/222718/files#r1147390358

**Comment by NickCao** [comment]
> It definitely does not work on my box without that override

I mean the override is just a hack instead of a proper solution.

> I gather we intentionally avoid the wrapper here?

But why, if it depends on qt, it is a qt application. The wrapper provides the correct QT_PLUGIN_PATH for the wayland backend (or other qt plugins) to be found.

**Comment by matklad** [comment]
Ah, that makes sense. I must say I am out of my depth here, so let cc @spacefrogg here :-)

**Review by AsPulse — CHANGES_REQUESTED** [review]
Hi, I'm author of  below pull requests that started the conversation, and appereciate your awesome suggestion!!
 - #291033

<br />

However, there is a slight deviation from the original intent, which I would be happy if you consider.
Original problem which I thought is that vivaldi is not working on gtk4 **even though it has `--gtk-version=4` as a option.** 
Vivaldi cannot found gtk4 manually installed by user, the only way is to include gtk4 as a dependency on the nixpkgs side.

<br />

I think we have two options:

1. So, If we decide that **the Nixpkgs side specify which gtk to use...**
    I believe that `--gtk-version-4` should be automatically added to default flags. if gtk4 selected.


2. Or as it has always been, If we decide that **the User side specify which gtk to use by flags...**
    What we should have is like...
    ```
    enableGtk3 ? true
    enableGtk4 ? false
    enableQt5 ? true
    enableQt6 ? false
    ```
    or, 
    ```
    gtk ? [gtk3]
    qt ? [qt5]
    ```
    In short, the should be able to include multiple package because users may use `--gtk-version=3`, or they may use `--gtk-version-4`.


Personally, I think the latter is better, but I am not sure....

File: pkgs/applications/networking/browsers/vivaldi/default.nix
Comment (spacefrogg) [inline pkgs/applications/networking/browsers/vivaldi/default.nix:115]: Also add the following to accommodate gtk4 support
```
--add-flags --gtk-version=${lib.versions.major gtk.version}
```

File: pkgs/applications/networking/browsers/vivaldi/default.nix
Comment (spacefrogg) [inline pkgs/applications/networking/browsers/vivaldi/default.nix:13]: Document, how they are supposed to be used,

**Review by spacefrogg — CHANGES_REQUESTED** [review]
Thanks! Please see other comments.

**Comment by 66Leo66** [comment]
Just upgraded to KDE 6, got a broken vivaldi and finally managed to bring it up using `vivaldi --disable-features=AllowQt`.

What's the suggested local fix before this gets merged (if it eventually will be)? And btw, shouldn't we land a fix asap, even if it's hacky, so that users won't face a broken browser after updating? (At least that what my understanding of distribution packaging's job...)

**Comment by yswtrue** [comment]
After upgrade to plasma 6, it still can't start direct from start menu, but it can start in terminal

**Comment by padhia** [comment]
> Just upgraded to KDE 6, got a broken vivaldi and finally managed to bring it up using `vivaldi --disable-features=AllowQt`.

Thank you for providing this workaround! I held off upgrading to Plasma 6 just because of this regression! 

I realize there are too many packages to test, but browsers are some of the core and highly personalized software that people can't function without.

**Comment by 66Leo66** [comment]
> > Just upgraded to KDE 6, got a broken vivaldi and finally managed to bring it up using `vivaldi --disable-features=AllowQt`.
> 
> Thank you for providing this workaround! I held off upgrading to Plasma 6 just because of this regression!
> 
> I realize there are too many packages to test, but browsers are some of the core and highly personalized software that people can't function without.

I'm glad you find it helpful! And yes, really, it's frustrating when a browser is completely broken, and even more so when a fix PR remains inactive for a week.
Not to blame anyone though. The volume of nixpkgs issues and prs is overwhelming I guess.

**Comment by matklad** [comment]
@spacefrogg PTAL:

* added doc comments for qt and gtk versions
* added `--gtk-version=` flag

Still not sure whether we should go with the whole "wrapQtApp" thing. But I think this is already an improvements, and I personally would rather stop here, as I don't want to do a larger refactor here.

**Comment by matklad** [comment]
>and even more so when a fix PR remains inactive for a week.

The PR was blocked on me doing the requested changes. In this situation, it might be better to submit a counter-PR which is better and can be merged as is (provided that you understand what maintainers want, and can actually send a PR which can be merged without requiring further review).

**Comment by spacefrogg** [comment]
> Still not sure whether we should go with the whole "wrapQtApp" thing. But I think this is already an improvements, and I personally would rather stop here, as I don't want to do a larger refactor here.

As I understand it, vivaldi is neither a qt nor a GTK application. It supports both libraries to blend-in with respective window managers. I suppose, that is what the qt5 and qt6 shims are there for. Thus, I would refrain from using any qt- or GTK-specific wrapper until this cannot be avoided. It would likely mean splitting the package into definite qt- and GTK-enabled versions which the user would have to be consciously picking.

**Comment by spacefrogg** [comment]
@66Leo66 @yswtrue @padhia I would like you to test, whether the changes work on your side, picking any of the four variants of GTK3/4 and qt5/6.

**Comment by padhia** [comment]
> @66Leo66 @yswtrue @padhia I would like you to test, whether the changes work on your side, picking any of the four variants of GTK3/4 and qt5/6.

Not sure if I am doing something wrong, but vivalid still fails.
```sh
~❯ nix run github:nixos/nixpkgs#vivaldi --impure                                                                                          via 🐍 v3.11.8 (.venv) via ❄  impure (sfconn-env) 
Gtk-Message: 10:00:18.502: Failed to load module "appmenu-gtk-module"
qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

[0324/100018.525458:ERROR:elf_dynamic_array_reader.h(64)] tag not found
[0324/100018.525534:ERROR:elf_dynamic_array_reader.h(64)] tag not found
[0324/100018.525806:ERROR:process_memory_range.cc(75)] read out of range
```

I am sure if override `qt` with `qt6`, I can then get `vivaldi` to open, but I'd have expected `qt6` to be somehow selected magically at build time depending on the DE the user has selected -- or at the very least, instead of failing, the command should refuse to be built/run, if the configuration isn't supported.

**Comment by spacefrogg** [comment]
No magic. So, it looks like we need four packages, then: `vivaldi-gtk3-qt5`, `vivaldi-gtk4-qt5`, `vivaldi-gtk3-qt6` and `vivaldi-gtk4-qt6`, each of which is a simple `vivaldi.override { ... }` with the relevant selections for gtk and qt.

**Comment by matklad** [comment]
>So, it looks like we need four packages

I'd rather leave that for someone else to do in a follow up: that will change the "public interface" of things, and I am not qualified enough to do that kind of change (and not motivated enough at this time to become qualified :) )

**Comment by matklad** [comment]
Status: my understanding is that this is waiting for review to either get merged, or get closed as an incorrect approach.

**Comment by JRodez** [comment]
I confirm that for Plasma 6 the following does the trick :
```nix
cus_vivaldi = pkgs.vivaldi.overrideAttrs (oldAttrs: {
      dontWrapQtApps = false; # Why this is True by default ?
      dontPatchELF = true;
      nativeBuildInputs = oldAttrs.nativeBuildInputs ++ [ pkgs.kdePackages.wrapQtAppsHook ];
    });
```

I'm not a Qt/Nix expert but  why it's not wrapped by default ?

**Comment by yswtrue** [comment]
> I confirm that for Plasma 6 the following does the trick :
> 
> ```nix
> cus_vivaldi = pkgs.vivaldi.overrideAttrs (oldAttrs: {
>       dontWrapQtApps = false;
>       dontPatchELF = true;
>       nativeBuildInputs = oldAttrs.nativeBuildInputs ++ [ pkgs.kdePackages.wrapQtAppsHook ];
>     });
> ```

works finally

**Comment by Daholli** [comment]
> I confirm that for Plasma 6 the following does the trick :
> 
> ```nix
> cus_vivaldi = pkgs.vivaldi.overrideAttrs (oldAttrs: {
>       dontWrapQtApps = false;
>       dontPatchELF = true;
>       nativeBuildInputs = oldAttrs.nativeBuildInputs ++ [ pkgs.kdePackages.wrapQtAppsHook ];
>     });
> ```

for me this now lets me start it from terminal again, but the taskbar icon just does nothing after some short loading.
it worked with 

```nix
vivaldi.override { commandLineArgs = "--disable-features=AllowQt"; };
```

for quite some time but after the last update it just stopped working. using the wrapQtAppsHook i can at least open it in the terminal now, but i cant figure out what is still wrong with the icon alone

**Comment by Daholli** [comment]
After some more testing, there are multiple versions in the nix store (due to me recompiling multiple times), that run out of the box, but the one linked in my .desktop file never does, anyone has some idea how i can properly update the desktop item, since the vivaldi package does not seem to use the makeDesktopIcon and it seems a bit backwards to reimplement the whole Install phase thing

**Comment by max06** [comment]
@Daholli somehow your options don't work for me

```
[flo@monster:~/.config/vivaldi]$ vivaldi --enable-features=UseOzonePlatform --ozone-platform=wayland --ozone-platform-hint=auto --enable-features=WaylandWindowDecorations 
qt.qpa.plugin: Could not find the Qt platform plugin "wayland" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, xcb.

[0708/172115.819678:ERROR:process_memory_range.cc(75)] read out of range
Aborted (core dumped)
``` 

Installed vivaldi with 

```
      ( vivaldi.overrideAttrs (oldAttrs: {
        dontWrapQtApps = false;
        dontPatchELF = true;
        nativeBuildInputs = oldAttrs.nativeBuildInputs ++ [ pkgs.kdePackages.wrapQtAppsHook ];
      }) )
``` 

(New to nixos, it could be me doing stupid things)

**Comment by Daholli** [comment]
You might also need the package qt5.qtwayland

**Comment by BonewheelMaster** [comment]
I was able to get Vivaldi running on Plasma 6 by installing the package `libsForQt5.qt5.qtwayland` and using the Wayland Ozone backend (achieved by adding `--ozone-platform=wayland`).

**Comment by Cybolic** [comment]
While I could get Vivaldi to run with adding `libsForQt5.qt5.qtwayland` and running it with `--ozone-platform=wayland` as well as this mentioned setup:

> I confirm that for Plasma 6 the following does the trick :
> 
> ```nix
> cus_vivaldi = pkgs.vivaldi.overrideAttrs (oldAttrs: {
>       dontWrapQtApps = false;
>       dontPatchELF = true;
>       nativeBuildInputs = oldAttrs.nativeBuildInputs ++ [ pkgs.kdePackages.wrapQtAppsHook ];
>     });
> ```

\- both of those setups produced a mix of blank Vivaldi windows and/or blank/non-rendering tabs.

What finally worked for me, was using the same change as is in this PR, which still works on the current version in nixpkgs unstable:

```nix
(pkgs.vivaldi.overrideAttrs (oldAttrs: {
  buildPhase = builtins.replaceStrings
    ["for f in libGLESv2.so libqt5_shim.so ; do"]
    ["for f in libGLESv2.so libqt5_shim.so libqt6_shim.so ; do"]
    oldAttrs.buildPhase
  ;
})).override {
  qt5 = pkgs.qt6;
  commandLineArgs = [ "--ozone-platform=wayland" ];
  # The following two are just my preference, feel free to leave them out
  proprietaryCodecs = true;
  enableWidevine = true;
};
```

I'm running Plasma 6 on Wayland, so that might have something to do with my issues with qt5.

**Pull request closed without merging**
